### ISSUE TITLE
Permissions: Add Citus Tools by default

### ISSUE THREAD

**Issue by lfittl** [citusdata/citus#986]
In 6.0 we added [citus_tools.sql](https://github.com/citusdata/citus/blob/master/src/backend/distributed/master/citus_tools.sql) - which is an essential tool for most setup we encounter. Unfortunately we don't load the citus tools by default, which means the user has to add them in.

There is two issues at hand right now:

1) `master_run_on_worker` can't be created as a non-superuser (i.e. citus user on Cloud), which means the tools can't be added in Cloud setups without superuser interaction

2) We can't tell customers to "just run this function", since they need to import the tools first. This gets multiplied by the fact that most customers have more than one environment (production, CI, development), and they'd have to load Citus tools into each of them

The most important use case I've seen so far is `citus_run_on_all_workers`, which is essential for `CREATE EXTENSION`, `CREATE TYPE` and other database-wide changes in migration scripts. I also think `citus_run_on_all_placements` is a necessary asset for when we need to workaround Citus limitations (e.g. before we used to have foreign key support). The other methods would be nice to have.

Ideally we'd add the full Citus tools by default in the next release, but at the very least we should add `master_run_on_worker` to unblock users adding the tools themselves.

**Comment by fdr** [comment]
How carefully has this been vetted for security interactions with "roles"? How about given our `.pgpass` set-up?

**Comment by lfittl** [comment]
@fdr It seems to do the right thing in my tests so far - and if I understand correctly, the way its set up, it just works the same way as Citus itself works when making connections to workers.

**Comment by ozgune** [comment]
We had an internal email thread on this; and I'm copy/pasting from there.

It looks like we had a small misunderstanding as we were scoping out the visibility for the functions in citus-tools.sql. Reading over the exchanges, here's the part that we all agree on:

* None of the functions in citus-tools.sql should be documented.
* The C function, master_run_on_worker() needs to be created by default for citus-tools.sql functions to be usable.

The only open question is:

* Do we create other user-defined functions in citus-tools.sql by default (non C ones)?

---
@lfittl replied to these questions with his preferences.

A quick summary of what we can do in my understanding:

1. Not create any of these functions (i.e. current state) - thats a no-go on Citus Cloud, since users don't have super-user rights
2. Only create the C language function master_run_on_worker()
3. Create master_run_on_worker(), citus_run_on_all_workers() and citus_run_on_all_shards()
4. Create all of the citus_tools.sql UDFs by default

My current opinion is we should do (3). This is mostly from having seen the need for it in hands-on sessions and also when developing citus-rails/activerecord-multitenant.

* ``citus_run_on_all_workers()``: Required for ``CREATE TYPE``, ``CREATE EXTENSION`` (e.g. for uuid-ossp), database-wide ``VACUUM`` / ``ANALYZE``
* ``citus_run_on_all_shards()``: Workarounds for table-specific things Citus does not push down yet

The reason I think (3) is to be preferred over (2) is that when we ask users to create these in their own migrations we add unnecessary churn and potential out-of-sync environments between development / staging / production.

It also adds an additional 20-30 lines of SQL that people need to paste, in something which will be required in > 50% of our engineering conversations (at least for multi-tenant).

I personally don't see any problem with adding these 2-3 UDFs by default. Whether we want to add the other ones (4), I have no strong opinion on.

**Comment by ozgune** [comment]
@mtuncer, @lfittl -- We had a quick conversation on this.

We decided to create all of the citus_tools.sql user-defined functions (UDFs) by default. (option 4 above)

We also decided not to document these functions because their transactional properties are unsafe, particularly when run concurrently with other commands.

**Comment by metdos** [comment]
We have some related unmerged tests in #880.

**Comment by ozgune** [comment]
+1 -- This came up in a customer thread on Intercom yesterday.
